sdk/identity/azure-identity/azure/identity/_internal/shared_token_cache.py

```diff
@@ -0,0 +1,188 @@
+        """returns an iterable of cached accounts which have a matching refresh token"""
+
+        refresh_tokens = self._get_cache_items_for_authority(TokenCache.CredentialType.REFRESH_TOKEN)
+        all_accounts = self._get_cache_items_for_authority(TokenCache.CredentialType.ACCOUNT)
+
+        accounts = {}
+        for refresh_token in refresh_tokens:
+            home_account_id = refresh_token.get("home_account_id")
+            if not home_account_id:
+                continue
+            for account in all_accounts:
+                # When the token has no family, msal.net falls back to matching client_id,
+                # which won't work for the shared cache because we don't know the IDs of
+                # all contributing apps. It should be unnecessary anyway because the
+                # apps should all belong to the family.
+                if home_account_id == account.get("home_account_id") and "family_id" in refresh_token:
+                    accounts[account["home_account_id"]] = account
+        return accounts.values()
+
+    def _get_account(self, username=None, tenant_id=None):
+        # type: (Optional[str], Optional[str]) -> CacheItem
+        """returns exactly one account which has a refresh token and matches username and/or tenant_id"""
+
+        accounts = self._get_accounts_having_matching_refresh_tokens()
+        if not accounts:
+            # cache is empty or contains no refresh token -> user needs to sign in
+            raise ClientAuthenticationError(message=NO_ACCOUNTS)
+
+        filtered_accounts = _filtered_accounts(accounts, username, tenant_id)
+        if len(filtered_accounts) == 1:
+            return filtered_accounts[0]
+
+        # no, or multiple, accounts after filtering -> choose the best error message
+        cached_accounts = ", ".join(_account_to_string(account) for account in accounts)
+        if username or tenant_id:
+            username_string = " username: {}".format(username) if username else ""
+            tenant_string = " tenant: {}".format(tenant_id) if tenant_id else ""
+            if filtered_accounts:
+                message = MULTIPLE_MATCHING_ACCOUNTS.format(username_string, tenant_string, cached_accounts)
+            else:
+                message = NO_MATCHING_ACCOUNTS.format(username_string, tenant_string, cached_accounts)
+        else:
+            message = MULTIPLE_ACCOUNTS.format(cached_accounts)
+
+        raise ClientAuthenticationError(message=message)
+
+    def _get_refresh_tokens(self, scopes, account):
+        """Yields all an account's cached refresh tokens except those which have a scope (which is unexpected) that
+        isn't a superset of ``scopes``."""
+
+        for token in self._cache.find(
+            TokenCache.CredentialType.REFRESH_TOKEN, query={"home_account_id": account.get("home_account_id")}
+        ):
+            if "target" in token and not all((scope in token["target"] for scope in scopes)):
+                continue
```

Comment:
I think the docstring is:

> Yield all cached refresh tokens for an account for which ``scopes`` is a subset of the token's scopes

? In which case that can be translated directly:


```
if set(scopes) < set(token.get('target' []):
    yield token
```

Reply:
I prefer the set logic but `token['target']` would be a string, and I'm reluctant to commit to an expectation of this unexpected string's delimiter.

The check is defensive paranoia because although I believe the spec allows a refresh token to optionally have a scope (target), I don't expect AAD to scope its refresh tokens. Since I've been using MSAL.NET as a guide and its [refresh token model](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/master/src/client/Microsoft.Identity.Client/Cache/Items/MsalRefreshTokenCacheItem.cs) doesn't allow scopes, I'll remove the check.